File: sdks/python/src/ice/handler/handler.py
"""Handler implementation for Ice SDK."""

from __future__ import annotations

from enum import IntFlag
from typing import TYPE_CHECKING

from ice.enums import TimeType
from ice._internal.timeutil import time_disabled
from ice import log

if TYPE_CHECKING:
    from ice.context.roam import Roam
    from ice.node.base import Node


class DebugFlag(IntFlag):
    """Debug flags for handler."""
    IN_ROAM = 1
    PROCESS = 2
    OUT_ROAM = 4


class Handler:
    """
    Handler manages the execution of an ice rule tree.

    Corresponds to Java's IceHandler.
    """

    def __init__(self) -> None:
        self.ice_id: int = 0
        self.conf_id: int = 0
        self.scenes: set[str] = set()
        self.time_type: TimeType = TimeType.NONE
        self.start: int = 0
        self.end: int = 0
        self.debug: int = 0
        self.root: Node | None = None

    def has_debug(self, flag: DebugFlag) -> bool:
        """Check if a debug flag is set."""
        return (self.debug & flag) != 0

    def _meta_suffix(self, roam: Roam) -> dict:
        """Build log suffix from meta: id/scene/nid/ts, only include non-zero values."""
        s: dict = {}
        if roam.get_id() > 0:
            s["id"] = roam.get_id()
        scene = roam.get_scene()
        if scene:
            s["scene"] = scene
        if roam.get_nid() > 0:
            s["nid"] = roam.get_nid()
        s["ts"] = roam.get_ts()
        return s

    def handle(self, roam: Roam) -> None:
        if self.root is None:
            return

        if time_disabled(self.time_type, roam.get_ts(), self.start, self.end):
            return

        from ice.log import set_trace_id, reset_trace_id
        token = set_trace_id(roam.get_trace())
        try:
            if self.has_debug(DebugFlag.IN_ROAM):
                log.info("handle input", extra={"roam": roam, **self._meta_suffix(roam)})

            self.root.process(roam)

            if self.has_debug(DebugFlag.PROCESS):
                log.info("handle process", extra={"process": roam.get_process_info(), **self._meta_suffix(roam)})

            if self.has_debug(DebugFlag.OUT_ROAM):
                log.info("handle output", extra={"roam": roam, **self._meta_suffix(roam)})

        except Exception as e:
            log.error("handler error", extra={"error": str(e), **self._meta_suffix(roam)})
            raise
        finally:
            reset_trace_id(token)

    def handle_with_node_id(self, roam: Roam) -> None:
        if self.root is None:
            return

        from ice.log import set_trace_id, reset_trace_id
        token = set_trace_id(roam.get_trace())
        try:
            if self.has_debug(DebugFlag.IN_ROAM):
                log.info("handle input", extra={"roam": roam, **self._meta_suffix(roam)})

            self.root.process(roam)

            if self.has_debug(DebugFlag.PROCESS):
                log.info("handle process", extra={"process": roam.get_process_info(), **self._meta_suffix(roam)})

            if self.has_debug(DebugFlag.OUT_ROAM):
                log.info("handle output", extra={"roam": roam, **self._meta_suffix(roam)})

        except Exception as e:
            log.error("handler error", extra={"error": str(e), **self._meta_suffix(roam)})
            raise
        finally:
            reset_trace_id(token)

    def __repr__(self) -> str:
        return f"Handler(ice_id={self.ice_id}, scenes={self.scenes})"

File: sdks/go/handler/handler.go
// Package handler provides the ice handler implementation.
package handler

import (
	"encoding/json"
	stdctx "context"

	icecontext "github.com/zjn-zjn/ice/sdks/go/context"
	"github.com/zjn-zjn/ice/sdks/go/enum"
	"github.com/zjn-zjn/ice/sdks/go/internal/timeutil"
	"github.com/zjn-zjn/ice/sdks/go/log"
	"github.com/zjn-zjn/ice/sdks/go/node"
)

// DebugFlag constants for controlling debug output.
const (
	DebugInRoam  byte = 1 << 0 // Log input roam
	DebugProcess byte = 1 << 1 // Log execution process
	DebugOutRoam byte = 1 << 2 // Log output roam
)

// Handler represents an ice handler that processes requests.
type Handler struct {
	IceId    int64
	ConfId   int64
	Scenes   map[string]struct{}
	TimeType enum.TimeType
	Start    int64
	End      int64
	Debug    byte
	Root     node.Node
}

// NewHandler creates a new Handler.
func NewHandler() *Handler {
	return &Handler{
		Scenes: make(map[string]struct{}),
	}
}

// metaSuffix builds log args from meta: id/scene/nid/ts, only non-zero values.
func metaSuffix(roam *icecontext.Roam) []any {
	var args []any
	if roam.GetId() > 0 {
		args = append(args, "id", roam.GetId())
	}
	if roam.GetScene() != "" {
		args = append(args, "scene", roam.GetScene())
	}
	if roam.GetNid() > 0 {
		args = append(args, "nid", roam.GetNid())
	}
	args = append(args, "ts", roam.GetTs())
	return args
}

// roamWithoutIce returns JSON string of roam data excluding _ice.
func roamWithoutIce(roam *icecontext.Roam) string {
	data := roam.Data()
	delete(data, "_ice")
	b, _ := json.Marshal(data)
	return string(b)
}

// Handle processes a roam.
func (h *Handler) Handle(ctx stdctx.Context, roam *icecontext.Roam) {
	if timeutil.TimeDisabled(h.TimeType, roam.GetTs(), h.Start, h.End) {
		return
	}

	if h.hasDebug(DebugInRoam) {
		args := append([]any{"roam", roamWithoutIce(roam)}, metaSuffix(roam)...)
		log.Info(ctx, "handle in", args...)
	}

	if h.Root != nil {
		h.Root.Process(ctx, roam)

		if h.hasDebug(DebugProcess) {
			args := append([]any{"process", roam.GetProcess().String()}, metaSuffix(roam)...)
			log.Info(ctx, "handle process", args...)
		}
		if h.hasDebug(DebugOutRoam) {
			args := append([]any{"roam", roamWithoutIce(roam)}, metaSuffix(roam)...)
			log.Info(ctx, "handle out", args...)
		}
	} else {
		log.Error(ctx, "root not exist", metaSuffix(roam)...)
	}
}

// HandleWithNodeId processes a roam using node ID.
func (h *Handler) HandleWithNodeId(ctx stdctx.Context, roam *icecontext.Roam) {
	if h.hasDebug(DebugInRoam) {
		args := append([]any{"roam", roamWithoutIce(roam)}, metaSuffix(roam)...)
		log.Info(ctx, "handle in", args...)
	}

	if h.Root != nil {
		h.Root.Process(ctx, roam)

		if h.hasDebug(DebugProcess) {
			args := append([]any{"process", roam.GetProcess().String()}, metaSuffix(roam)...)
			log.Info(ctx, "handle process", args...)
		}
		if h.hasDebug(DebugOutRoam) {
			args := append([]any{"roam", roamWithoutIce(roam)}, metaSuffix(roam)...)
			log.Info(ctx, "handle out", args...)
		}
	}
}

// SetScenes sets the scenes for this handler.
func (h *Handler) SetScenes(scenes []string) {
	h.Scenes = make(map[string]struct{}, len(scenes))
	for _, s := range scenes {
		h.Scenes[s] = struct{}{}
	}
}

// HasScene checks if the handler has the given scene.
func (h *Handler) HasScene(scene string) bool {
	_, ok := h.Scenes[scene]
	return ok
}

func (h *Handler) hasDebug(flag byte) bool {
	return (h.Debug & flag) != 0
}
